onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,719 @@
+{
+  bool modified = false;
+  ORT_ENFORCE(node != nullptr);
+  if (IsCastTo(*node, TensorProto::FLOAT16)) {
+    std::unordered_set<NodeArg*> require_cast;
+    NodeArg* cast_input = node->MutableInputDefs()[0];
+    std::unordered_set<NodeArg*> require_type_change;
+    SearchUpstream(graph, cast_input, require_cast, require_type_change, removed_nodes, level);
+    if (require_cast.size() > 0 && require_cast.find(cast_input) == require_cast.end()) {
+      // Remove Cast operation
+      VLOGS(logger, 1) << "PropagateBackwards: Removed Cast node  " << node->Name();
+      RemoveCastNodes(graph, {node}, removed_nodes);
+      InsertCastNodes(graph, require_cast, true, removed_nodes);
+      ChangeTypeToFP16(require_type_change, logger);
+      VLOGS(logger, 1) << "PropagateBackwards: Inserted Cast nodes "
+                << ConcatNames<std::unordered_set<NodeArg*>>(require_cast);
+      VLOGS(logger, 1) << "PropagateBackwards: Changed the type from float to float16 : "
+                << ConcatNames<std::unordered_set<NodeArg*>>(require_type_change);
+      modified = true;
+    }
+  } else if (node->OpType() != "Cast") {
+    for (NodeArg* input: node->MutableInputDefs()) {
+      Node* producer = graph.GetMutableProducerNode(input->Name());
+      if (producer != nullptr &&
+          std::find(removed_nodes.begin(), removed_nodes.end(), producer->Index()) == removed_nodes.end()) {
+         modified |= PropagateBackwards(graph, producer, removed_nodes, level, logger);
+      }
+    }
+  }
+  return modified;
+}
+
+// Fuse all nodes, replace with a single node.
+// Assumptions:
+// 1. all nodes are Cast ops and are of the same Cast type
+// 2. all the nodes have the same input
+static void FuseNodes(Graph& graph, NodeArg* input, std::vector<Node*> nodes,
+                                    std::deque<onnxruntime::NodeIndex>& removed_nodes)
+{
+  std::vector<NodeArg*> outputs;
+  for (Node* node : nodes) {
+    std::vector<NodeArg*> node_outputs = node->MutableOutputDefs();
+    outputs.insert(outputs.end(), node_outputs.begin(), node_outputs.end());
+  }
+  Node* node = nodes[0];
+  (void) graph.AddNode(graph.GenerateNodeName(node->Name() + "_replace"),
+                       node->OpType(),
+                       "Created to replace a node",
+                       {input},
+                       outputs,
+                       &node->GetAttributes(),
+                       node->Domain());
+  for (Node* n : nodes) {
+    removed_nodes.push_back(n->Index());
+    graph_utils::RemoveNodeOutputEdges(graph, *n);
+  }
+}
+
+// Traverse the graph recursively searching/collecting sibling Cast op nodes to fuse and call FuseNodes.
+static bool FuseSiblingCasts(Graph& graph, Node* parent,
+                          std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                          const logging::Logger& logger)
+{
+  bool modified = false;
+  for (NodeArg* output : parent->MutableOutputDefs()) {
```

Comment:
To go over the consumer of a node, you can use 
for (auto it = node->OutputNodesBegin(); it != node->OutputNodesEnd(); ++it)

Reply:
I think you are asking using the nested loop:
  for (NodeArg* output : parent->MutableOutputDefs()) {
  ...
    for (Node* node : graph.GetMutableConsumerNodes(output->Name())) {
    ...
    }
 }

In this case I am only iterating over all the consumer nodes the given output and I am using the output in the body of the loop.